File: Sources/GatewayCore/AudioProfile.swift
import Foundation

/// Listening levels, kept between launches.
///
/// **These are not render settings.** Nothing here changes a generated wav, so
/// nothing here belongs in `VoiceProfile.renderKey` — turning the pink noise
/// down must never invalidate a cached take. That separation is the same one
/// the voice profile already draws between what renders and what only gates.
///
/// Hand-editable JSON with `decodeIfPresent` throughout, so a file written
/// before a field existed still loads, and a dropped key falls back instead of
/// emptying the mix.
public struct AudioProfile: Codable, Sendable, Equatable {
    /// The narration. Its own level, because the voice is the content and the
    /// bed is the room.
    public var speech: Double = 1.0
    /// Retained open-mouth vocalisation. Independent of the bed master so a
    /// quiet binaural/noise calibration does not make the human hum vanish.
    public var resonantTuning: Double = 0.50
    /// The retained wake-up signal is intentionally more assertive. It remains
    /// adjustable rather than baking loudness into the source recording.
    public var returnSignal: Double = 0.85
    /// The binaural pair — carrier and differential.
    public var hemiSync: Double = 0.45
    public var pinkNoise: Double = 0.35
    public var whiteNoise: Double = 0.0
    /// Surf and the other session-level textures.
    public var surf: Double = 0.30
    /// Master, applied after everything else.
    public var master: Double = 0.8

    public init() {}

    public static let range = 0.0...1.0

    /// Every level, in the order the panel shows them.
    public var levels: [(name: String, value: Double)] {
        [("speech", speech), ("resonant tuning", resonantTuning),
         ("return signal", returnSignal), ("hemi-sync", hemiSync),
         ("pink noise", pinkNoise), ("white noise", whiteNoise),
         ("surf", surf), ("bed master", master)]
    }

    public func retainedMediaLevel(for role: AudioAssetRole) -> Double {
        switch role {
        case .resonantTuning: resonantTuning
        case .returnSignal: returnSignal
        }
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        speech = try c.decodeIfPresent(Double.self, forKey: .speech) ?? 1.0
        resonantTuning = try c.decodeIfPresent(Double.self, forKey: .resonantTuning) ?? 0.50
        returnSignal = try c.decodeIfPresent(Double.self, forKey: .returnSignal) ?? 0.85
        hemiSync = try c.decodeIfPresent(Double.self, forKey: .hemiSync) ?? 0.45
        pinkNoise = try c.decodeIfPresent(Double.self, forKey: .pinkNoise) ?? 0.35
        whiteNoise = try c.decodeIfPresent(Double.self, forKey: .whiteNoise) ?? 0.0
        surf = try c.decodeIfPresent(Double.self, forKey: .surf) ?? 0.30
        master = try c.decodeIfPresent(Double.self, forKey: .master) ?? 0.8
    }

    /// Clamped on the way in, so a hand-edited 11 does not blow the mix.
    public var clamped: AudioProfile {
        var p = self
        func c(_ v: Double) -> Double { min(max(v, 0), 1) }
        p.speech = c(speech); p.resonantTuning = c(resonantTuning)
        p.returnSignal = c(returnSignal); p.hemiSync = c(hemiSync)
        p.pinkNoise = c(pinkNoise)
        p.whiteNoise = c(whiteNoise); p.surf = c(surf); p.master = c(master)
        return p
    }
}

public enum AudioProfileIO {
    /// One file, beside the other things the app remembers about the user.
    public static func url(root: URL) -> URL { root.appending(path: "memory/audio.json") }

    public static func load(root: URL) -> AudioProfile {
        guard let d = try? Data(contentsOf: url(root: root)),
              let p = try? JSONDecoder().decode(AudioProfile.self, from: d)
        else { return AudioProfile() }
        return p.clamped
    }

    public static func save(_ profile: AudioProfile, root: URL) throws {
        let u = url(root: root)
        try FileManager.default.createDirectory(at: u.deletingLastPathComponent(),
                                                withIntermediateDirectories: true)
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        try enc.encode(profile.clamped).write(to: u, options: .atomic)
    }
}

File: cross-platform/src/core/audioProfile.ts
/**
 * The listening levels, ported from `Sources/GatewayCore/AudioProfile.swift`.
 *
 * A headphone calibration, not a mix: what the bed's parts should sound like
 * *to this person on these headphones*, saved once and applied on top of
 * whatever the tape itself asks for. The bed is unlistenable without it —
 * the engine's own constants are raw source levels, and `master` alone is the
 * difference between a signal and a shout.
 *
 * Deliberately free of any file access so the audio worklet can import it: a
 * worklet global scope has no module resolver for `fs`, and one `import` of it
 * anywhere in this file's graph would take the whole bed engine down with it.
 * Reading and writing the file lives in `audioProfileStore.ts`, the same split
 * Swift makes between `AudioProfile` and `AudioProfileIO`.
 */

export interface AudioProfile {
  /** The narration. Its own level, because the voice is the content and the
   *  bed is the room. */
  speech: number;
  /** Retained open-mouth vocalisation. Independent of the bed master so a
   *  quiet binaural/noise calibration does not make the human hum vanish. */
  resonantTuning: number;
  /** The retained wake-up signal is intentionally more assertive. It remains
   *  adjustable rather than baking loudness into the source recording. */
  returnSignal: number;
  /** The binaural pair — carrier and differential. */
  hemiSync: number;
  pinkNoise: number;
  whiteNoise: number;
  /** Surf and the other session-level textures. */
  surf: number;
  /** Master, applied after everything else. */
  master: number;
}

/** Swift's memberwise defaults, which are also what every absent key in the
 *  saved file falls back to. */
export const defaultAudioProfile = (): AudioProfile => ({
  speech: 1.0,
  resonantTuning: 0.50,
  returnSignal: 0.85,
  hemiSync: 0.45,
  pinkNoise: 0.35,
  whiteNoise: 0.0,
  surf: 0.30,
  master: 0.8,
});

export const audioProfileRange = { min: 0, max: 1 } as const;

const clamp01 = (v: number): number => Math.min(Math.max(v, 0), 1);

/** Clamped on the way in, so a hand-edited 11 does not blow the mix. */
export function clampedAudioProfile(p: AudioProfile): AudioProfile {
  return {
    speech: clamp01(p.speech),
    resonantTuning: clamp01(p.resonantTuning),
    returnSignal: clamp01(p.returnSignal),
    hemiSync: clamp01(p.hemiSync),
    pinkNoise: clamp01(p.pinkNoise),
    whiteNoise: clamp01(p.whiteNoise),
    surf: clamp01(p.surf),
    master: clamp01(p.master),
  };
}

/**
 * Swift's `init(from:)`: every field is `decodeIfPresent … ?? default`, so a
 * partial or hand-edited file loses only the keys it actually omits rather
 * than resetting the whole calibration. Note it does *not* clamp — loading
 * does, which is the same order the original keeps.
 */
export function decodeAudioProfile(raw: unknown): AudioProfile {
  const d = defaultAudioProfile();
  if (typeof raw !== "object" || raw === null) return d;
  const o = raw as Record<string, unknown>;
  const num = (key: keyof AudioProfile): number => {
    const v = o[key];
    return typeof v === "number" && Number.isFinite(v) ? v : d[key];
  };
  return {
    speech: num("speech"),
    resonantTuning: num("resonantTuning"),
    returnSignal: num("returnSignal"),
    hemiSync: num("hemiSync"),
    pinkNoise: num("pinkNoise"),
    whiteNoise: num("whiteNoise"),
    surf: num("surf"),
    master: num("master"),
  };
}

/** Every level, in the order the panel shows them. */
export function audioProfileLevels(p: AudioProfile): { name: string; value: number }[] {
  return [
    { name: "speech", value: p.speech },
    { name: "resonant tuning", value: p.resonantTuning },
    { name: "return signal", value: p.returnSignal },
    { name: "hemi-sync", value: p.hemiSync },
    { name: "pink noise", value: p.pinkNoise },
    { name: "white noise", value: p.whiteNoise },
    { name: "surf", value: p.surf },
    { name: "bed master", value: p.master },
  ];
}
